File: console_ui.py
# -*- coding: utf-8 -*-
"""
控制台单行动态显示工具
提供：单行刷新、进度条、旋转动画、yt-dlp/ffmpeg 进度解析
"""

import re
import sys
import time
import math
from threading import Thread, Event, Lock

# ── 终端单行刷新锁（B2）──────────────────────────────────────────────────────
# 并发转码时多个任务会同时刷新同一行 stderr，无锁会导致转义序列交错、显示乱码。
# 所有单行写都必须走这把锁，保证每次刷新以完整的一行落盘。
_term_lock = Lock()


# ── 旋转动画帧 ───────────────────────────────────────────────────────────────────
SPINNER = ['⠋', '⠙', '⠹', '⠸', '⠼', '⠴', '⠦', '⠧', '⠇', '⠏']


# ── 单行刷新（覆盖当前行）────────────────────────────────────────────────────────
def update_line(text: str) -> None:
    """\x1b[2K = 清除整行, \r = 回到行首"""
    with _term_lock:
        sys.stderr.write(f"\x1b[2K\r{text}")
        sys.stderr.flush()


def clear_line() -> None:
    """清除当前行"""
    with _term_lock:
        sys.stderr.write('\x1b[2K\r')
        sys.stderr.flush()


# ── 格式化工具 ───────────────────────────────────────────────────────────────────
def fmt_size(bytes_val: float) -> str:
    if bytes_val < 1024:
        return f"{bytes_val:.0f}B"
    if bytes_val < 1024 * 1024:
        return f"{bytes_val / 1024:.1f}KiB"
    if bytes_val < 1024 * 1024 * 1024:
        return f"{bytes_val / 1024 / 1024:.1f}MiB"
    return f"{bytes_val / 1024 / 1024 / 1024:.1f}GiB"


def fmt_time(sec: float) -> str:
    if not math.isfinite(sec) or sec < 0:
        return '--:--'
    m = int(sec) // 60
    s = int(sec) % 60
    return f"{m:02d}:{s:02d}"


def fmt_speed(bps: float) -> str:
    if not bps or not math.isfinite(bps):
        return '---'
    if bps < 1024:
        return f"{bps:.0f}B/s"
    if bps < 1024 * 1024:
        return f"{bps / 1024:.1f}KiB/s"
    return f"{bps / 1024 / 1024:.1f}MiB/s"


# ── 文本进度条 ───────────────────────────────────────────────────────────────────
def text_bar(percent: float, width: int = 18) -> str:
    filled = round(percent / 100 * width)
    empty = width - filled
    return f"[{'█' * filled}{'░' * max(0, empty)}]"


# ── 旋转动画 Spinner ───────────────────────────────────────────────────────────
class Spinner:
    """线程安全的旋转动画。
    用法:
        spinner = Spinner()
        spinner.start("处理中")
        # ... 执行耗时操作 ...
        spinner.stop("处理完成")
    """

    def __init__(self):
        self._thread: Thread | None = None
        self._stop_event = Event()
        self._label = ''
        self._start_time = 0.0
        self._final_text = ''

    def start(self, label: str) -> None:
        self.stop()
        self._label = label
        self._start_time = time.monotonic()
        self._stop_event.clear()
        self._final_text = ''

        def _spin():
            frame = 0
            while not self._stop_event.is_set():
                elapsed = int(time.monotonic() - self._start_time)
                s = f"  {SPINNER[frame % len(SPINNER)]} {self._label}... {elapsed}s"
                update_line(s)
                frame += 1
                self._stop_event.wait(0.16)

        self._thread = Thread(target=_spin, daemon=True)
        self._thread.start()

    def stop(self, final_text: str = '') -> None:
        if self._thread and self._thread.is_alive():
            self._stop_event.set()
            self._thread.join(timeout=1)
            self._thread = None
        if final_text:
            with _term_lock:
                sys.stderr.write('\x1b[2K\r')
                sys.stderr.write(f"  {final_text}\n")
                sys.stderr.flush()


# ── yt-dlp 进度行解析 ────────────────────────────────────────────────────────
# 输入示例：
#   [download]  45.2% of  100.00MiB at  12.34MiB/s ETA 00:04
#   [download] Destination: xxx.f248.mp4
#   [download] Merging formats into "xxx.mp4"
# 返回：dict 或 None
def parse_ytdlp_line(line: str) -> dict | None:
    """解析 yt-dlp 输出行，返回结构化的进度信息或 None"""
    if '[download] Destination:' in line:
        m = re.search(r'Destination:.*\.(mp4|webm|mkv|flv|f\d+)', line)
        if m:
            return {'type': 'dest', 'ext': m.group(1)}
    if 'Merging formats into' in line:
        return {'type': 'merge'}
    # 解析进度：允许 ~ 前缀和 ETA 格式
    m = re.search(
        r'\[download\]\s+([\d.]+)%\s+of\s+~?\s*([\d.]+)(MiB|KiB|GiB)\s+at\s+([\d.]+)(MiB|KiB|GiB)/s(?:\s+ETA\s+(\d+:\d+))?',
        line,
    )
    if m:
        return {
            'type': 'progress',
            'percent': float(m.group(1)),
            'downloaded': float(m.group(2)),
            'downloaded_unit': m.group(3),
            'speed': float(m.group(4)),
            'speed_unit': m.group(5),
            'eta': m.group(6) or '',
        }
    return None


# ── ffmpeg 进度解析（需配合 -progress pipe:1 -nostats）────────────────
# 输入为 key=value 行，如：out_time_us=12345678
_ffmpeg_state = {'duration_us': 0, 'out_time_us': 0, 'speed': 0, 'total_size': 0}


def parse_ffmpeg_progress(line: str, total_duration_sec: float = 0, state: dict | None = None) -> dict | None:
    """解析 ffmpeg -progress pipe:1 输出行，返回进度对象或 None。

    state: 进度累积状态字典。传入独立字典即可让每个转码任务拥有自己的进度，
    避免并发转码时共享 _ffmpeg_state 导致百分比串扰/跳 0（B1）。
    不传则回退到模块级 _ffmpeg_state（向后兼容）。
    """
    if state is None:
        state = _ffmpeg_state
    line = line.strip()
    if not line or line.startswith('['):
        return None
    m = re.match(r'^(\w+)=(.+)$', line)
    if not m:
        return None
    key, val = m.group(1), m.group(2).strip()
    if key == 'out_time_us':
        state['out_time_us'] = int(val) if val else 0
    elif key == 'speed':
        speed_m = re.match(r'([\d.]+)x', val)
        state['speed'] = float(speed_m.group(1)) if speed_m else 0
    elif key == 'total_size':
        state['total_size'] = int(val) if val else 0
    elif key == 'duration_us':
        state['duration_us'] = int(val) if val else 0
    # 计算进度
    dur = total_duration_sec * 1e6 if total_duration_sec > 0 else (state['duration_us'] or 1)
    percent = min(100, (state['out_time_us'] / dur) * 100)
    return {
        'type': 'progress',
        'percent': round(percent * 10) / 10,
        'elapsed': state['out_time_us'] / 1e6,
        'speed': state['speed'],
        'total_size': state['total_size'],
    }


def reset_ffmpeg_state() -> None:
    """重置 ffmpeg 进度状态"""
    _ffmpeg_state.update({'duration_us': 0, 'out_time_us': 0, 'speed': 0, 'total_size': 0})

File: console-ui.mjs
/**
 * 控制台单行动态显示工具
 * 提供：单行刷新、进度条、旋转动画、yt-dlp/fmpeg 进度解析
 */

// ── 旋转动画帧 ───────────────────────────────────────────────────────────────────
const SPINNER = ['⠋','⠙','⠹','⠸','⠼','⠴','⠦','⠧','⠇','⠏'];
let _spinTimer = null;
let _spinLabel = '';
let _spinStart = 0;

// ── 单行刷新（覆盖当前行）────────────────────────────────────────────────────────
function updateLine(text) {
  // \x1b[2K = 清除整行, \r = 回到行首
  process.stderr.write(`\x1b[2K\r${text}`);
}

function clearLine() {
  process.stderr.write('\x1b[2K\r');
}

// ── 格式化工具 ───────────────────────────────────────────────────────────────────
function fmtSize(bytes) {
  if (bytes < 1024) return `${bytes}B`;
  if (bytes < 1024 * 1024) return `${(bytes / 1024).toFixed(1)}KiB`;
  if (bytes < 1024 * 1024 * 1024) return `${(bytes / 1024 / 1024).toFixed(1)}MiB`;
  return `${(bytes / 1024 / 1024 / 1024).toFixed(1)}GiB`;
}

function fmtTime(sec) {
  if (!isFinite(sec) || sec < 0) return '--:--';
  const m = Math.floor(sec / 60);
  const s = Math.floor(sec % 60);
  return `${String(m).padStart(2, '0')}:${String(s).padStart(2, '0')}`;
}

function fmtSpeed(bps) {
  if (!bps || !isFinite(bps)) return '---';
  if (bps < 1024) return `${bps.toFixed(0)}B/s`;
  if (bps < 1024 * 1024) return `${(bps / 1024).toFixed(1)}KiB/s`;
  return `${(bps / 1024 / 1024).toFixed(1)}MiB/s`;
}

// ── 文本进度条（无 ANSI 依赖，纯文本）────────────────────────────────────
function textBar(percent, width = 18) {
  const filled = Math.round(percent / 100 * width);
  const empty = width - filled;
  return `[${'█'.repeat(filled)}${'░'.repeat(Math.max(0, empty))}]`;
}

// ── 旋转动画 ───────────────────────────────────────────────────────────────────
function startSpinner(label) {
  stopSpinner();
  _spinLabel = label;
  _spinStart = Date.now();
  let frame = 0;
  _spinTimer = setInterval(() => {
    const elapsed = ((Date.now() - _spinStart) / 1000).toFixed(0);
    updateLine(`  ${SPINNER[frame % SPINNER.length]} ${_spinLabel}... ${elapsed}s`);
    frame++;
  }, 160);
}

function stopSpinner(finalText = '') {
  if (_spinTimer) {
    clearInterval(_spinTimer);
    _spinTimer = null;
  }
  if (finalText) {
    clearLine();
    process.stderr.write(`  ${finalText}\n`);
  }
}

// ── yt-dlp 进度行解析 ────────────────────────────────────────────────────────
// 输入示例：
//   [download]  45.2% of  100.00MiB at  12.34MiB/s ETA 00:04
//   [download] Destination: xxx.f248.mp4
//   [download] Merging formats into "xxx.mp4"
// 返回：{ percent, downloaded, total, speed, eta } 或 null
function parseYtdlpLine(line) {
  // 检测正在下载的流类型
  if (line.includes('[download] Destination:')) {
    const m = line.match(/Destination:.*\.(mp4|webm|mkv|flv|f\d+)/);
    if (m) {
      return { type: 'dest', ext: m[1] };
    }
  }
  if (line.includes('Merging formats into')) {
    return { type: 'merge' };
  }
  // 解析进度：允许千位分隔符空格和 ETA 格式
  const prog = line.match(
    /\[download\]\s+([\d.]+)%\s+of\s+~?\s*([\d.]+)(MiB|KiB|GiB)\s+at\s+([\d.]+)(MiB|KiB|GiB)\/s(?:\s+ETA\s+(\d+:\d+))?/
  );
  if (prog) {
    return {
      type: 'progress',
      percent: parseFloat(prog[1]),
      downloaded: parseFloat(prog[2]),
      downloadedUnit: prog[3],
      speed: parseFloat(prog[4]),
      speedUnit: prog[5],
      eta: prog[6] || '',
    };
  }
  return null;
}

// ── fmpeg 进度解析（需配合 -progress pipe:1 -nostats）────────────────
// 输入为 key=value 行，如：out_time_us=12345678\ntotal_size=456789\nspeed=1.5x
// 累积行直到出现 packet=1 或空行，然后返回进度对象
// 返回：{ percent, elapsed, speed } 或 null
let _ffmpegState = { durationUs: 0, outTimeUs: 0, speed: 0, totalSize: 0 };

function parseFfmpegProgress(line, totalDurationSec, state) {
  if (!state) state = _ffmpegState;  // 向后兼容：不传则用全局态
  line = line.trim();
  if (!line || line.startsWith('[')) return null;
  const kv = line.match(/^(\w+)=(.+)$/);
  if (!kv) return null;
  const key = kv[1];
  const val = kv[2].trim();
  if (key === 'out_time_us') {
    state.outTimeUs = parseInt(val, 10) || 0;
  } else if (key === 'speed') {
    const m = val.match(/([\d.]+)x/);
    state.speed = m ? parseFloat(m[1]) : 0;
  } else if (key === 'total_size') {
    state.totalSize = parseInt(val, 10) || 0;
  } else if (key === 'duration_us') {
    state.durationUs = parseInt(val, 10) || 0;
  }
  // 用 out_time_us / duration 计算进度
  const dur = totalDurationSec > 0 ? totalDurationSec * 1e6 : (state.durationUs || 1);
  const percent = Math.min(100, (state.outTimeUs / dur) * 100);
  return {
    type: 'progress',
    percent: Math.round(percent * 10) / 10,
    elapsed: state.outTimeUs / 1e6,
    speed: state.speed,
    totalSize: state.totalSize,
  };
}

// 每个转码任务创建独立的进度状态（B1），避免并发转码时百分比串扰/跳 0。
function makeFfmpegProgressParser(totalDurationSec) {
  const state = { durationUs: 0, outTimeUs: 0, speed: 0, totalSize: 0 };
  return (line) => parseFfmpegProgress(line, totalDurationSec, state);
}

function resetFfmpegState() {
  _ffmpegState = { durationUs: 0, outTimeUs: 0, speed: 0, totalSize: 0 };
}

// ── 导出 ──────────────────────────────────────────────────────────────────────────
export {
  updateLine,
  clearLine,
  fmtSize,
  fmtTime,
  fmtSpeed,
  textBar,
  startSpinner,
  stopSpinner,
  parseYtdlpLine,
  parseFfmpegProgress,
  makeFfmpegProgressParser,
  resetFfmpegState,
};
